### PR TITLE
Add telemetry to API

### PR DESCRIPTION
**Pull request by mike0sv**
Adds telemetry event logging to `mlem.api` commands. Some Qs:

1) Should we log API event if it was used by CLI command (in this PR we do not) 
2) What if one API command calls some nested API command? Should both events be logged, or only topmost? (kinda the same as 1) if you really think about it)
3) What parameters we want to log from each command (will post suggestions as a separate comment)
4) Should `save, load, load_meta` be added too? It can show interesting stuff like what types of models and datasets are used more frequently
5) Need a framework to enforce same extra fields for similar API/CLI calls, but that depends on previous Qs

### CODE REVIEW

**Comment by mike0sv** [comment]
General suggestion: wherever we use `MlemABC`, log it's alias (string type) and classpath (like `mlem.contrib.smth.ClassName`)

- apply: model type, data type, is batch used
- apply_remoe: client, data type
- clone: ???
- init: ???
- link: ???
- build: builder, model type
- serve: server, model type
- import_object: import hook type
- deploy: deployment type

🤔 : We can also get fs protocol used wherever path's are used (eg, to know that people clone from github to local, or build directly from/to s3)

**Comment by mike0sv** [comment]
cc: @shcheklein @omesser @aguschin

**Comment by shcheklein** [comment]
Hey, are there examples of libraries that call home? (even CLI raises some concerns, library can give us even more troubles).

Second question - are we wrapping it into a separate thread?

**Comment by mike0sv** [comment]
From the top of my head  I think boto3 logs every request, not sure tbh but seen something like this when I debugged something

We spawn a separate process even

**Comment by mike0sv** [comment]
Don't see any difference between API and CLI tbh. Most of the concerns I heard are about PI and sending it to someone who sells the data (like GA). Since we use our own open source tools to collect telemetry we should be good, anybody can see what fields we are collecting if they have doubts

**Comment by omesser** [comment]
Some thoughts:
> 1. Should we log API event if it was used by CLI command (in this PR we do not)

What is the reason for telemetry here? if it's to show usage / adoption - then I would treat API and CLI as competing entry points. meaning, if someone ran a command via CLI, not register api event, assuming we are tracking api event in order to learn about the direct usage of API


> What if one API command calls some nested API command? Should both events be logged, or only topmost? (kinda the same as 1) if you really think about it)

Same, I would think why are we logging this. if it's to show usage than we probably shouldn't

> 3. What parameters we want to log from each command (will post suggestions as a separate comment)

Suggest to consult with engineers from DVC and CML about what they chose to register in their telemetry - usually rule of thumb should be to white-list specific stuff (model architecture, deployment technology) , being very mindful to collect only a handful of needed bits of information only and nothing that could accidentally be looked at as personal/sensitive information like file names, paths, env variables etc.

> 4. Should save, load, load_meta be added too? It can show interesting stuff like what types of models and datasets are used more frequently

👍 

> Second question - are we wrapping it into a separate thread?

+1 should not effect behavior or have super minimal/negligable effect (functionality or performance)

> even CLI raises some concerns, library can give us even more troubles

This is very true. We should probably put out some kind of blog post explaining what and when are we logging and explain so we can point concerned users there. Also if we're adding this to api it should be VERY explicit and easy to turn off (func signature)

**Comment by aguschin** [comment]
@mike0sv 

> 1. Should we log API event if it was used by CLI command (in this PR we do not)

No IMO, on't see a reason for this

> 2. What if one API command calls some nested API command? Should both events be logged, or only topmost? (kinda the same as 1) if you really think about it)

No, don't see a reason as well

> 3. What parameters we want to log from each command (will post suggestions as a separate comment)

Will update your comment suggestion if you don't mind.

> 4. Should save, load, load_meta be added too? It can show interesting stuff like what types of models and datasets are used more frequently

Yes!

@omesser 

> We should probably put out some kind of blog post explaining what and when are we logging and explain so we can point concerned users there.

Good idea!

> Also if we're adding this to api it should be VERY explicit and easy to turn off (func signature)

Wouldn't turning it off in MLEM settings be enough? So it's either
1. editing `.mlem.yaml` manually or by `mlem config set ...`
2. doing that on-the-fly in Python runtime with something like
```
import mlem.config
mlem.config.analytics = False
```

Having it in each func signature could be a non-convenient overkill IMO.

**Comment by skshetry** [comment]
> Don't see any difference between API and CLI tbh.

Environment matters. It should not do what's asked, definitely not phone home.

Also, at this stage of `mlem`([with ~800 downloads/month](https://pypistats.org/packages/mlem)), what insights do we think this will provide us?

**Comment by aguschin** [comment]
> what insights do we think this will provide us

I would expect some insights about:
1. Which ML frameworks and Data formats people use (and which they _try_ to use, but MLEM fails) - this is what we can get from `mlem.api.save()`. Here we can pay more attention to some frameworks or prioritize support new ones.
2. Is MLEM used programmatically at all - and how frequently. Right now, I didn't pay attention to API at all (except for `save/load`) TBH. I assume CLI commands should be few times widely used. If I happen to be wrong, I'll take closer look at API to see how people us it, etc.

I guess we can extract other things listed above from CLI usage as well. But at least these 2 can't be extracted without API analytics I assume.

**Comment by mike0sv** [comment]
I find it kinda ironic that we implement tracking stuff in PR #451 🤣

**Comment by mike0sv** [comment]
I've created https://github.com/iterative/telemetry-python/pull/49 based on code in this PR. Once it's merged I will use it to update code in this PR

**Comment by omesser** [comment]
@aguschin 
> Wouldn't turning it off in MLEM settings be enough?

Yeah, for CLI it makes sense 👍 . but for code usage - I'm afraid it might not be clear. 
I totally get @shcheklein and @skshetry concerns here - even if it's not that clear cut. I think we should take them very seriously and come off as super friendly to the user.
I'm thinking that making it explicit in api signatures might be a bit dirty, yes, but easy enough so users won't feel like it's too sneaky, because you step into the code, you see the kwarg, it's obvious. If you import a package and use api, you don't think about adding a config and editing it, it's not obvious, to me at least 🤔 so may be perceived as sneaky/non-trivial to turn-off and that's not friendly.

So it's your choice folks, but yeah, let's be very careful to make it obvious and easy to turn off, and it should be very safe, never harm perf, or fail the execution (separate thread?)

**Comment by mike0sv** [comment]
You can turn it off with `mlem.telemetry.telemetry.enabled = False`

I think we should write a blogpost that explains how, what and why we collect.

Maybe even add a link to it in api module docstring? No sure about function docstrings - to much duplication.

Also when we reach 1.0 we can gradually remove some reports, or maybe disable API reporting by default. But it makes sense that we want as much feedback as possible while MLEM is not mature

**Comment by mike0sv** [comment]
basically blocked by https://github.com/iterative/mlem/pull/472

**Review by aguschin — APPROVED** [review]

File: mlem/core/metadata.py
Comment (aguschin) [reply mlem/core/metadata.py:175]: Decided to skip it for now.

**Pull request merged**
